awscli/botocore/utils.py

```diff
@@ -3305,6 +3339,231 @@ def fetch_token(
+        return botocore.awsrequest.AWSResponse(None, 200, {}, None), {}
+
+    def _get_base_authorization_uri(self):
+        """Simulates an SSO-OIDC request so that we can extract the "base"
+        endpoint for the current client to use for the un-modeled Authorize
+        operation
+        """
+        self._client.meta.events.register('before-call',
+                                          self._extract_resolved_endpoint)
+        self._client.register_client(
+            clientName='temp',
+            clientType='public'
+        )
+        self._client.meta.events.unregister('before-call',
+                                            self._extract_resolved_endpoint)
+
+        return self._base_endpoint
+
+    def _get_authorization_uri(
+            self,
+            client_id,
+            registration_scopes,
+            expected_state):
+
+        query_params = {
+            'response_type': 'code',
+            'client_id': client_id,
+            'redirect_uri': self._auth_code_fetcher.redirect_uri_with_port(),
+            'state': expected_state,
+            'code_challenge_method': 'S256'
+            # Don't want to encode code_challenge again, so we append below
+        }
+
+        # For the query param, scopes must be space separated before encoding
+        if registration_scopes:
+            query_params['scope'] = " ".join(registration_scopes)
+
+        return (
+            f'{self._get_base_authorization_uri()}/authorize?'
+            f'{urllib.parse.urlencode(query_params)}'
+            f'&code_challenge={self.code_challenge[:-1]}'  # trim final '='
+        )
+
+    def _get_new_token(self, start_url, session_name, registration_scopes):
+        registration = self._registration(
+            start_url,
+            session_name,
+            registration_scopes,
+        )
+
+        expected_state = uuid.uuid4()
+
+        authorization_uri = self._get_authorization_uri(
+            registration['clientId'],
+            registration_scopes,
+            expected_state)
+
+        # Even though there's just one uri, this matches the inputs
+        # for the device code flow so that we can reuse the browser handlers
+        authorization_args = {
+            'verificationUri': authorization_uri,
+            'verificationUriComplete': authorization_uri,
+            'userCode': None
+        }
+
+        # Open/display the link, then block until the redirect uri is hit and
+        # the auth code is retrieved
+        self._on_pending_authorization(**authorization_args)
+        auth_code, state = self._auth_code_fetcher.get_auth_code_and_state()
+
+        if auth_code is None:
+            raise AuthorizationCodeLoadError(
+                error_msg='Failed to retrieve an authorization code.'
+            )
+
+        if state != expected_state:
+            raise AuthorizationCodeLoadError(
+                error_msg='State parameter does not match expected value.'
```

Comment:
It is opaque because it's generated internally. This is guarding against CSFR, in case an an attacker is able to hit the redirect endpoint with _their_ authorization code, injecting the attackers credentials into the user's application/workflow.
* https://datatracker.ietf.org/doc/html/rfc6749#section-10.12

Users shouldn't see it short of a bug, bit flip, or if something is hitting the redirect before the normal worfklow does.

Do you think it's worth adding something like _"Try again"_?

VS Code as a reference: https://github.com/aws/aws-toolkit-vscode/blob/3b6197b2cfa8345a0775096f06e0983aec50102b/packages/core/src/auth/sso/server.ts#L34-L38